File: Aman_Bhaiya_DSA_List/Arrays/LQ27.java
package Aman_Bhaiya_DSA_List.Arrays;

import java.util.*;

public class LQ27 {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<>();
        int n = nums.length;

        for(int i = 0; i < n-2; i++){
            if(i == 0 || (i > 0 && nums[i]!=nums[i-1])){
                int l = i+1;
                int r = n-1;
                int sum = 0 - nums[i];
                while(l < r){
                    if(nums[l] + nums[r] == sum){
                        ArrayList<Integer> arr = new ArrayList<>(
                            Arrays.asList(nums[i],nums[l],nums[r])
                        );
                        ans.add(arr);

                        while(l < r && nums[l] == nums[l+1]) l++;
                        while(l < r && nums[r] == nums[r-1]) r--;

                        l++;
                        r--;
                    }
                    else if(nums[l]+nums[r] < sum) l++;
                    else r--;
                }
            }
        }

        return ans;
    }
}

File: Aman_Bhaiya_DSA_List/Arrays/LQ27.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ans = []
        n = len(nums)
        nums.sort()

        for i in range(0,n-1):
            if i == 0 or (i > 0 and nums[i] != nums[i-1]):
                l = i+1
                r = n-1
                sum = 0 - nums[i]

                while l < r:
                    if nums[l] + nums[r] == sum:
                        lst = [nums[i],nums[l],nums[r]]
                        ans.append(lst)

                        while l < r and nums[l] == nums[l+1]:
                            l = l + 1
                        while l < r and nums[r] == nums[r-1]: 
                            r = r - 1
                        l = l + 1
                        r = r - 1
                    elif nums[l] + nums[r] < sum:
                        l = l + 1
                    else:
                        r = r - 1

        return ans
